File: ios/Sources/LiveRelaySDK/Publisher.swift
import Foundation
import AVFoundation
import WebRTC

/// Publishes the local camera + microphone (or screen) to the LiveRelay SFU.
///
/// Mirrors the reference web flow (`liverelay.js` → `LiveRelay.publish`):
/// capture → ICE servers → PeerConnection → addTrack → tuned offer →
/// POST /sfu/publish → set remote answer.
public final class Publisher {

    // MARK: - Public API (contract)

    public weak var delegate: LiveRelaySessionDelegate?

    public var capture: MediaCapture { _capture }

    public init(config: LiveRelayConfig, room: String) {
        self.room = room
        self.signaling = SignalingClient(config: config)
        self._capture = MediaCapture(factory: PeerConnectionProvider.shared.factory)
    }

    public func start(screen: Bool = false) async throws {
        notify(.connecting)
        do {
            // 1. Audio session for voice chat (speaker + bluetooth).
            try AudioSessionConfigurator.configureForVoiceChat()

            // 2. Local media (defaults: front camera, 1280x720@30 + mic).
            try await _capture.startCamera(position: .front, width: 1280, height: 720, fps: 30)
            _capture.startMicrophone()

            // 3. ICE servers + PeerConnection with internal state delegate.
            let iceServers = try await signaling.fetchIceServers()
            let pcDelegate = PCDelegate(owner: self)
            self.pcDelegate = pcDelegate
            let pc = PeerConnectionProvider.shared.makePeerConnection(
                iceServers: iceServers,
                delegate: pcDelegate
            )
            self.pc = pc

            // 4. Add tracks, then tune for low latency (before createOffer).
            if let videoTrack = _capture.videoTrack {
                let sender = pc.add(videoTrack, streamIds: ["stream"])
                if let videoTransceiver = pc.transceivers.first(where: { $0.mediaType == .video }) {
                    LatencyTuning.preferH264(on: videoTransceiver)
                }
                if let sender {
                    LatencyTuning.maintainFramerate(on: sender)
                }
            }
            if let audioTrack = _capture.audioTrack {
                _ = pc.add(audioTrack, streamIds: ["stream"])
            }

            // 5. Tuned offer → publish → remote answer.
            let offer = try await PeerConnectionProvider.shared.makeTunedOffer(for: pc)
            let answer = try await signaling.publish(offer: offer, room: room, screen: screen)
            try await PeerConnectionProvider.shared.setRemoteAnswer(answer, on: pc)
        } catch {
            teardown()
            notify(.failed)
            throw error
        }
    }

    public func stop() {
        teardown()
        notify(.closed)
    }

    // MARK: - Internals

    private let room: String
    private let signaling: SignalingClient
    private let _capture: MediaCapture
    private var pc: RTCPeerConnection?
    private var pcDelegate: PCDelegate?

    private let stateLock = NSLock()
    private var lastState: SessionState = .new

    private func teardown() {
        _capture.stop()
        pcDelegate?.invalidate()
        pcDelegate = nil
        pc?.close()
        pc = nil
    }

    /// Forwards a state change to the session delegate (deduplicated, on main).
    fileprivate func notify(_ state: SessionState) {
        stateLock.lock()
        guard state != lastState else {
            stateLock.unlock()
            return
        }
        lastState = state
        stateLock.unlock()

        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.delegate?.session(self, didChangeState: state)
        }
    }

    // MARK: - PeerConnection delegate

    /// Internal RTCPeerConnectionDelegate: maps connection states to
    /// `SessionState` and forwards them to the owning Publisher.
    private final class PCDelegate: NSObject, RTCPeerConnectionDelegate {

        private weak var owner: Publisher?

        init(owner: Publisher) {
            self.owner = owner
            super.init()
        }

        func invalidate() {
            owner = nil
        }

        // Unified connection state — the primary source for SessionState.
        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didChange newState: RTCPeerConnectionState) {
            let mapped: SessionState
            switch newState {
            case .new: mapped = .new
            case .connecting: mapped = .connecting
            case .connected: mapped = .connected
            case .disconnected: mapped = .disconnected
            case .failed: mapped = .failed
            case .closed: mapped = .closed
            @unknown default: mapped = .disconnected
            }
            owner?.notify(mapped)
        }

        // Required delegate methods (publisher sends only — mostly no-ops).
        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didChange stateChanged: RTCSignalingState) {}

        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didAdd stream: RTCMediaStream) {}

        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didRemove stream: RTCMediaStream) {}

        func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {}

        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didChange newState: RTCIceConnectionState) {}

        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didChange newState: RTCIceGatheringState) {}

        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didGenerate candidate: RTCIceCandidate) {
            // ICE candidates are gathered inline by makeTunedOffer (no trickle).
        }

        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didRemove candidates: [RTCIceCandidate]) {}

        func peerConnection(_ peerConnection: RTCPeerConnection,
                            didOpen dataChannel: RTCDataChannel) {}
    }
}

File: android/liverelay-sdk/src/main/java/app/orizn/liverelay/Publisher.kt
package app.orizn.liverelay

import android.content.Context
import org.webrtc.CandidatePairChangeEvent
import org.webrtc.DataChannel
import org.webrtc.IceCandidate
import org.webrtc.IceCandidateErrorEvent
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.RtpReceiver
import org.webrtc.RtpTransceiver

/**
 * Publishes the local camera + microphone (or screen) to a LiveRelay room
 * via POST /sfu/publish, mirroring the `LiveRelay.publish()` flow of
 * liverelay.js v0.5.0.
 *
 * Note: [LiveRelayListener] callbacks are invoked on internal WebRTC threads;
 * callers must post to their own thread/handler if needed.
 */
class Publisher(
    private val context: Context,
    private val config: LiveRelayConfig,
    private val room: String
) {
    var listener: LiveRelayListener? = null

    val capture: MediaCapture = MediaCapture(context)

    private val signaling = SignalingClient(config)

    @Volatile
    private var peerConnection: PeerConnection? = null

    @Volatile
    private var state: SessionState = SessionState.NEW

    private fun setState(next: SessionState) {
        if (state == next) return
        state = next
        listener?.onStateChanged(next)
    }

    /**
     * Internal observer: maps PeerConnection state to [SessionState] and
     * forwards it to [listener]. All other callbacks are clean no-ops
     * (a publisher neither receives remote tracks nor trickles ICE —
     * candidates are bundled into the offer by makeTunedOffer).
     */
    private val observer = object : PeerConnection.Observer {
        override fun onConnectionChange(newState: PeerConnection.PeerConnectionState) {
            when (newState) {
                PeerConnection.PeerConnectionState.NEW -> Unit // keep current state
                PeerConnection.PeerConnectionState.CONNECTING -> setState(SessionState.CONNECTING)
                PeerConnection.PeerConnectionState.CONNECTED -> setState(SessionState.CONNECTED)
                PeerConnection.PeerConnectionState.DISCONNECTED -> setState(SessionState.DISCONNECTED)
                PeerConnection.PeerConnectionState.FAILED -> setState(SessionState.FAILED)
                PeerConnection.PeerConnectionState.CLOSED -> {
                    // stop() reports CLOSED itself; only relay unexpected closes.
                    if (state != SessionState.CLOSED) setState(SessionState.DISCONNECTED)
                }
            }
        }

        override fun onSignalingChange(newState: PeerConnection.SignalingState?) = Unit
        override fun onIceConnectionChange(newState: PeerConnection.IceConnectionState?) = Unit
        override fun onIceConnectionReceivingChange(receiving: Boolean) = Unit
        override fun onIceGatheringChange(newState: PeerConnection.IceGatheringState?) = Unit
        override fun onIceCandidate(candidate: IceCandidate?) = Unit
        override fun onIceCandidateError(event: IceCandidateErrorEvent?) = Unit
        override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>?) = Unit
        override fun onSelectedCandidatePairChanged(event: CandidatePairChangeEvent?) = Unit
        override fun onAddStream(stream: MediaStream?) = Unit
        override fun onRemoveStream(stream: MediaStream?) = Unit
        override fun onDataChannel(channel: DataChannel?) = Unit
        override fun onRenegotiationNeeded() = Unit
        override fun onAddTrack(receiver: RtpReceiver?, streams: Array<out MediaStream>?) = Unit
        override fun onRemoveTrack(receiver: RtpReceiver?) = Unit
        override fun onTrack(transceiver: RtpTransceiver?) = Unit
    }

    /**
     * Starts publishing: configures audio, captures camera + mic, performs
     * the SDP offer/answer exchange with the SFU.
     *
     * @param screen forwarded to the server so it routes video to screen_tx.
     * @throws LiveRelayException on signaling failure.
     */
    suspend fun start(screen: Boolean = false) {
        setState(SessionState.CONNECTING)
        try {
            // 1. Audio routing for low-latency communication.
            AudioConfigurator.configureForCommunication(context)

            // 2. Local media.
            PeerConnectionProvider.initialize(context)
            capture.startCamera()
            capture.startMicrophone()
            val videoTrack = capture.videoTrack
                ?: throw LiveRelayException("Camera capture failed: no video track")
            val audioTrack = capture.audioTrack
                ?: throw LiveRelayException("Microphone capture failed: no audio track")

            // 3. PeerConnection with server-provided ICE config.
            val iceServers = signaling.fetchIceServers()
            val pc = PeerConnectionProvider.makePeerConnection(iceServers, observer)
            peerConnection = pc

            // 4. Send-only tracks, framerate kept under constraint.
            val videoSender = pc.addTrack(videoTrack, listOf("stream"))
            pc.addTrack(audioTrack, listOf("stream"))
            LatencyTuning.maintainFramerate(videoSender)

            // 5. Offer (Opus tuned + H264 preferred + ICE bundled) -> publish -> answer.
            val offer = PeerConnectionProvider.makeTunedOffer(pc)
            val answer = signaling.publish(offer, room, screen)
            PeerConnectionProvider.setRemoteAnswer(pc, answer)
        } catch (e: Exception) {
            setState(SessionState.FAILED)
            releaseResources()
            throw e
        }
    }

    /** Stops capture, closes the PeerConnection and restores audio routing. */
    fun stop() {
        releaseResources()
        setState(SessionState.CLOSED)
    }

    private fun releaseResources() {
        // Fermer la PeerConnection AVANT de disposer les tracks capturés :
        // disposer un track encore attaché à un sender actif est un usage natif invalide.
        peerConnection?.let { pc ->
            try {
                pc.close()
            } finally {
                pc.dispose()
            }
        }
        peerConnection = null
        capture.stop()
        AudioConfigurator.reset(context)
    }
}
